---
title: "Report on SmithWaterman package"
author: "Maria Sonechkina"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Report on SmithWaterman package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(SmithWaterman)
```

## Main aim of the package

The main goal of this package is to perform a local pairwise alignment between two nucleotide sequences using the Smith-Waterman algorithm.

## History and motivation

In the field of molecular sequence analysis, an important problem has been the identification of homologous subsequences given a set of long sequences. This task of sequences alignment is used also for non-biological sequences, but in bionformatics this procedure is performed in order to identify some regions of similarity that could be the consequence of functional, structural or evolutionary relationships between the sequences.  
So, in this case, the task addresses nucleotide sequences, which are the main type of sequence for which it is looked for the homology with another nucleotide sequence. 

However, the issue that arises with nucleotide sequence is that it is necessary considering the presence of deletions or insertions, which are typical variations of the genomic DNA. 

The first developments began with the Needleman & Wunsch algoirthm (1970), an heuristic homology algorithm which first introduced an iterative matrix method for computation and it is widely used for pairwise global sequence alignment tasks.
After that, it has been developed a true metric measure of the distance between sequences (Sellers, 1974), that has been generalized by Waterman _et al._ (1976) to take into account the deletions/insertions of any length. This metric measures the minimum number of "mutational events" required to convert one sequence into another.

The Smith-Waterman algorithm extends the original idea of the homology algorithm of Needleman & Wunsch and aims to find a pair of segments, one from each of the two long sequences, such that there is no other pair of segments with greater similarity. This procedure is different than the Needleman-Wunsch algorithm since it finds the best local alignment between two subsequences with respect to the scoring system being used, instead of the global one.

Another difference from the Needleman & Wunsch algorithm is that the Smith-Waterman algorithm compares segments of all possible lengths and optimizes the similarity measure, instead of lookung at the entire sequence.

The main motivation to use this local pairwise alignment algorithm is that it is difficult to obtain correct alignments in regions of low similarity between distantly related sequences, since mutation has added noise over evolutionary time to allow meaningful comparison of those regions.

...

## Development process

The premise for the developmental process is that Smith-Waterman algorithm aligns two sequences by considering:
- matches,
- mismatches, which represent the substitutions in the  biological sequence,
- gaps, which are introduced by insertions and deletions in the biological sequences.

### Scoring system and substitution matrix
The first idea of the function is to take as input two string sequences and three attributes that correspond to the scores for matches, mismatches and gaps. The general idea is to assign positive scores to the matches and lower ones to the mismatches. The gap penalty determines the score cost for the presence of gaps.

The inputs are processed in two ways:
1. the scores are checked to be numerical type, otherwise the algorithm cannot be initiated;
2. the two strings are arranged to be the row and the column of the scoring matrix

## Use case

### References

Smith, Temple F. & Waterman, Michael S. (1981). "Identification of Common Molecular
Subsequences". Journal of Molecular Biology. 147 (1): 195â€“197.
